### ISSUE TITLE
String handling does not conform to Lua 5.3

### ISSUE THREAD

**Issue by mjanicek** [mjanicek/rembulan#2]
Lua strings in Rembulan are currently mapped to `java.lang.String`. While this is convenient for Java interoperability and is in line with the way Lua booleans and numbers are handled, it is not consistent with the definition on strings as per [§2.1 of the Lua Reference Manual](http://www.lua.org/manual/5.3/manual.html):

> The type *string* represents immutable sequences of bytes. Lua is 8-bit clean: strings can contain any 8-bit value, including embedded zeros ('\0'). Lua is also encoding-agnostic; it makes no assumptions about the contents of a string.

`java.lang.String`s are essentially immutable sequences of Java `char`, i.e., 16-bit signed integers. In order to treat `java.lang.String` as a Lua string, there ought to be a value mapping from 16-bit values to 8 bits (e.g. taking the least significant byte). 

### Problems

This is a (probably incomplete) list of reasons why using `java.lang.String` for Lua strings is broken:

 * Currently, string equality is determined using the `equals()` method. This means that regardless of how the 16-bit-to-8-bit value mapping is chosen, there will always be strings `a` and `b` such that `!a.equals(b)`, but in Lua, `a == b` in Lua. Additionally, it would be possible that `a.hashCode() != b.hashCode()` in Java, but `a == b` in Lua.
 * This means that, unless strings are "normalised", (i.e. converted using the 16-bit-to-8-bit value map to a new string wrapping the mapped values), using strings as keys in a `HashMap` may lead to a situation where the map contains multiple keys that are equal in Lua.
 * In order to be able to use `HashMap` in a way compatible with Lua equality, normalisation would need to be done before insertion *and* retrieval, which may seriously degrade performance. Furthermore, there is no way of telling whether a given instance of `java.lang.String` is normalised or not, so it would need to be done every time.
 * `java.lang.String` does not carry around information about its character encoding, so normalisation must be charset-independent. This could lead to very confusing behaviour. Custom normalisation methods could be provided, but they would not bring any benefits performance-wise (see the previous point).
 * There is a mismatch in the *spirit* of Lua strings and `java.lang.String`. While Lua strings do not care about their contents, Java strings do: the `char` type is explicitly meant to represent 16-bit Unicode characters.
 * 16-bit strings require twice as much space as 8-bit strings to be represented correctly (though the actual memory usage is JVM-dependent).

#### Keep `java.lang.String`

Some of the problems above may be avoided:

 * Lua string comparison can take the value mapping into account.
 * Lua table implementations can be implemented from scratch, without using `HashMap` or other collections that use `equals()` to determine the equality of keys.

However, there are still outstanding issues:

 * Strings equal in Lua may still be unequal in Java, and there would be no way of distinguishing them. In Lua, this is no problem, but it opens up an entire class of hard-to-detect bugs in the Java code that interoperates with Lua:
```lua
-- in Lua, it does not matter whether we return a or b... but it might in Java!
if a == b then return a end
```
 * The *spirit* problem still remains, and turns into a smell.

#### Use a custom immutable byte string class

The only correct solution appears to be to use another class (a "byte string") for representing Lua strings. This of course has its downsides:

 * The impact on the code base would be non-trivial.
 * There is no immutable byte string class in the JDK.
 * `java.lang.String` is ubiquitous in the Java world: introducing extra conversion step to and from the new class would make Java interoperability more cumbersome.
 * If the rest of the value mapping rules remain, `java.lang.String` would become *light userdata*, which may lead to confusion, especially if booleans are still mapped to `java.lang.Boolean` and numbers to `java.lang.Number`.
 * `java.lang.String` can be interned; it might be difficult to get such functionality for byte strings.
 * Similarly for [string deduplication](http://openjdk.java.net/jeps/192) (since Java 8 Update 20).

**Comment by bensku** [comment]
I'm no expert here, but wouldn't using custom String class make deduplication of them by JVM impossible? It is quite useful feature...

Also, I think for most use cases running Lua is due to using some Java code with Lua code, and wanting to transmit data between them without JNI/serialization overhead. Having different string type on Lua side would seriously do some harm for that

For example, I'm thinking that I might allow modding for (currently WIP) game using Java and Lua, latter with Rembulan. That would need the ability to pass all data between Java and Lua quickly - for example, if I add chat event listener in Lua, it should not be expensive to send string from Java side to it and get changed string back from Lua side. It could be really lots of those listeners if server admin install even few mods.

Tl;dr; I think going with Java strings is better idea, even if it doesn't conform with Lua spec.

**Comment by mjanicek** [comment]
That's a very good point -- I've added it to the list. This kind of shuffling between Java and Lua could indeed easily become a bottleneck (especially if those strings are big): like boxing/unboxing, but this time for strings.

**Comment by mjanicek** [comment]
Plus there's this (coming in Java 9): [JEP 254: Compact Strings](http://openjdk.java.net/jeps/254). `java.lang.String` is getting quite some attention, so abandoning it might not be a particularly good idea.

**Comment by mjanicek** [comment]
I'm exploring the following hybrid approach: do introduce a new `ByteString` class, but make it lazy. It comes in two flavours:
 * one backed by a `java.lang.String`, with a `Charset` attached to it;
 * the other backed by a `byte[]`.

The idea is to have `ByteString` behave like a byte sequence without doing too much unnecessary work. This means that we'd only like it to behave as a byte sequence when it is taking part in a Lua operation:
 * the length operator (`#`);
 * concatenation (`..`);
 * comparison (e.g., `<`);
 * used as a table key (`t[x]`);
 * used by standard library functions (mostly `string.*`, `io.*` and `utf8.*`).

In a `java.lang.String`-backed `ByteString`:

* **Length:** the byte length can be cached without having to keep the bytes around (just needs to be computed once).[1]

* **Concatenation:** when concatenating with another `java.lang.String`-backed `ByteString` that also uses the same `Charset`, reduce to a normal `java.lang.String` concatenation; otherwise, convert both operands to `byte[]` and concatenate.[2]

* **Comparison:** comparison with another `java.lang.String`-backed `ByteString` with the same `Charset` may *probably* (depending on the `Charset`!) reduce to a normal Java lexicographical comparison. Otherwise, convert to `byte[]` and compare.[1][2]

* **Table indices:** tables require `hashCode()` and `equals()` that use the byte sequence interpretation (as established in the original problem statement). `hashCode()` can be computed once and subsequently cached; `equals()` with a `java.lang.String`-backed `ByteString` with the same `Charset` reduces to a normal `java.lang.String.equals()`, otherwise both operands can be converted to `byte[]` and compared for equality.[1][2]

For `byte[]`-backed `ByteStrings`, the operations are similar. (Such a `ByteString` would be used by e.g. `string.pack` and `io.read`.)

[1] -- length, comparison, `hashCode()` and `equals()` may be done in "incremental" mode: we do not need to keep the actual byte representation around, but can do this character per character. Although this is certainly more complex and quite possibly slower than just getting a `byte[]`, it would still work for very big strings.

[2] -- the result of the conversion to `byte[]` (or better the `byte[]`-flavoured `ByteString`) may be cached behind a `SoftReference`. For standard library functions, this is probably a requirement: for instance, it would be *very* wrong to have `string.byte` convert the entire string to `byte[]` every time it is called... just to get a single byte out of it.

As a bonus, `ByteString`s could also be ropes. That would mostly pay off in concatenation- and serialisation-heavy contexts. (For example, an HTTP server.)

**Comment by bensku** [comment]
Is there going to be a way to automatically pass stringa between Java and Lua? Anyway, this is good idea!

**Comment by mjanicek** [comment]
> Is there going to be a way to automatically pass strings between Java and Lua?

Yes: one way to get a `ByteString` is to get "thin" wrapper of a `java.lang.String` without copying any data. Plus, all `java.lang.String`s are implicitly converted to `ByteString` this way in the runtime. Alternatively, the user can also pass a "real" `ByteString` to the runtime (one that internally uses a `byte` array).

I'm quite happy with this approach: I think this is the way to go. It is probably the best that can be done on a generic JVM in order to see strings as byte sequences, while avoiding the overhead of having to do `byte[]`<->`char[]` conversions on the Java/Lua boundary (most of the time).

**Comment by bensku** [comment]
Implicit conversion was the thing I needed to know. Thanks!

**Comment by mjanicek** [comment]
This has now been implemented and merged into master. There are still potential improvements to be made to the underlying representations (e.g., caching), but those can be made without changing the API.
